### PR TITLE
Configurable default _limit parameter

### PR DESCRIPTION
**Pull request by jozefcipa**
### What does it do?

This PR introduces a new options field in `config/server.json`. `defaultLimit` defines how many results should be returned if a request is called without an explicit `_limit` param, whereas `maxLimit` specifies the max allowed number of results that can be requested from an API (mainly due to performance) (See https://github.com/strapi/strapi/issues/4931#issuecomment-572522899)

```
rest: {
  defaultLimit: 50,
  maxLimit: 100,
}
```

### Why is it needed?
It allows developers to customize the default `_limit` value which is now set to 100. 

### Some considerations and questions
While working on this feature I run into some issues that I think are worth to reconsider.

1. I noticed there is an option to set `_limit` to **-1** in order to query all the results which I think is completely opposite of what this PR tries to accomplish, so I'd suggest to either get rid of it (don't know how important or used it is), or remove `maxLimit` option from this PR.
2. If the limit is empty but present (think of `?_limit=`), or is set to **0** it seems to work exactly as if it was set to **-1**. I'd suggest setting the `defaultLimit` instead.
3. If the limit is some non-integer or negative (other than -1) value it crashes with 500 error. Here I would again prefer returning a `defaultLimit`.
4. I'd also like to show some kind of pagination info in the response, what do you think? I do realize it would be a breaking change. https://github.com/strapi/strapi/issues/4931#issuecomment-587940593

P.S. Regarding the 500 error, I find it quite inappropriate to return 500 for such a small mistake. When browsing the code I didn't find some kind of "known errors" which are common to happen but wouldn't end up showing 500 to the user but rather some 400 Bad Request with a proper error response and description. ~~I can open a separate issue for this and possibly work on that if confirmed to be a good idea.~~ Wouldn't it be better to replace `throw new Error` with `strapi.errors.badRequest` so it can give some valuable feedback to users rather than just generic `Internal server error`? I noticed that pattern in multiple places so I'm wondering if it shouldn't be changed.

Please let me know what you think guys 🙏 

### Related issue
Resolves https://github.com/strapi/strapi/issues/4931

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
> I noticed there is an option to set _limit to -1 in order to query all the results which I think is completely opposite of what this PR tries to accomplish, so I'd suggest to either get rid of it (don't know how important or used it is), or remove maxLimit option from this PR.

We will want to maintain this functionality, in this case `-1` just means "all", though obviously we (Strapi) don't suggest it's usage as it can be quite performance heavy, there is still some users that do use it and we want to avoid a breaking change if possible. Obviously in the case of someone configuring a `maxLimit` the option should be disabled and either:

- Return the max limit
- Provide the user a readable 400 error (Simply stating the option is not allowed)

> If the limit is empty but present (think of ?_limit=), or is set to 0 it seems to work exactly as if it was set to -1. I'd suggest setting the defaultLimit instead.

I want to say the check is anything less than 1 means "request all" but will want the engineers to comment to see if they have a better idea.

> If the limit is some non-integer or negative (other than -1) value it crashes with 500 error. Here I would again prefer returning a defaultLimit.

If enabled I agree, either we return the limit or provide an error (maybe a configuration option for which one)

> I'd also like to show some kind of pagination info in the response, what do you think? I do realize it would be a breaking change. #4931 (comment)

We have discussed this before and I think we do plan to implement something, on my personal projects I created a koa middleware to add this information in a `content-range` header (total count based on the query parameters, not just the query itself) but I agree adding some kind of page information like `page 1 of x`. I believe @alexandrebodin has some plans for better paging options with our database refresh this quarter so I'll let him comment more.

---

Just a heads up the @strapi/maintainers may request you open an [RFC](https://github.com/strapi/rfcs) and approve it before we merge the PR as there is some considerations as to where we store the configuration information. Personally I think the `./config/api.js` that is used for a few other options is the best (and likewise any `./config/env/someenv/api.js` file too)

Just wanted to give you a warning beforehand.

**Comment by alexandrebodin** [comment]
Hi adding this implementation in the convert-rest-query-params is currently not the right place to do it as it is use in the db layer and would prevent query from call the db layer to get all the data. (this is something we plan on changing) For the time being you will need to add this logic to the core-api/service so it doesn't affect the Content Manager and the graphQL plugin that have there own limit management. 

Hope this helps. let me know if you need more guidance ;)

**Comment by jozefcipa** [comment]
@alexandrebodin I moved it to `strapi/core-strapi/service.js` as requested. It also checks if `maxLimit` was set and if so, it prevents using `-1` and returns `maxLimit` instead, as suggested by @derrickmehaffy. Please let me know what you think :)

File: packages/strapi/lib/core-api/service.js
Comment (derrickmehaffy) [inline packages/strapi/lib/core-api/service.js:18]: We should move these to the `./config/api.js` not the server.
https://strapi.io/documentation/v3.x/concepts/configurations.html#api

**Review by derrickmehaffy — CHANGES_REQUESTED** [review]
Only other thing is we need documentation for this.

File: packages/strapi/lib/core-api/service.js
Comment (jozefcipa) [reply packages/strapi/lib/core-api/service.js]: Config moved to `api.js` and added docs. Maybe it could be rephrased better, what do you think?

**Review by derrickmehaffy — APPROVED** [review]
LGTM but we still need the engineers to validate the variable names and expected response if someone exceeds the limit.

(Should we return only the max, or should we return an error)

File: packages/strapi/lib/core-api/service.js
Comment (alexandrebodin) [inline packages/strapi/lib/core-api/service.js:23]: even if it is an edge case if limit is 0 it will return the default but 0 is a valid limit :)

**Review by alexandrebodin — CHANGES_REQUESTED** [review]
I think we don't need to have the maxLimit set to -1 to specify unlimited. a null or undefiend maxLimit would work fine here :)

**Comment by alexandrebodin** [comment]
Hi @jozefcipa did you have time to continue working on this ? You will need to remove the doc and move it to the new documentation repo though :) @derrickmehaffy  can help you out on that

**Comment by jozefcipa** [comment]
@alexandrebodin nope, I didn't have much time and honestly forgot about it. I'll try to look at it in the nearest future :)

File: packages/strapi/lib/core-api/__tests__/service.test.js
Comment (jozefcipa) [inline packages/strapi/lib/core-api/__tests__/service.test.js:171]: @alexandrebodin I did some changes as requested, but as I was originally working on this task a few months ago, please read these test cases to make sure it works in the desired way 🙏

**Comment by jozefcipa** [comment]
@alexandrebodin I did some changes as suggested, please have a look 🙌 
Also, created a PR in the https://github.com/strapi/documentation/pull/55

File: packages/strapi/lib/core-api/__tests__/service.test.js
Comment (alexandrebodin) [inline packages/strapi/lib/core-api/__tests__/service.test.js:180]: can you pass maxLimit here instead of doing a special case ? and just setting it if defined in the input :)

Comment (jozefcipa) [reply packages/strapi/lib/core-api/__tests__/service.test.js:180]: yeah, I was thinking of that too, but I didn't like the idea much as it would require modification of the `getLimitParam` just for the test purposes 🤔 Also, I'm not sure how exactly `params` are passed from the request down here, but I'm afraid this way a user could do something like `/some-api-call?_limit=500&maxLimit=1000` and `maxLimit` would be accepted, or no? Correct me if I'm wrong.

Comment (alexandrebodin) [reply packages/strapi/lib/core-api/__tests__/service.test.js:180]: let me share want I meant :) Just make some logic in the test but don't change the code :)

```js
test.each([

])('Tes _limit %s', (input, expect) => {
  strapi.config.api.rest.maxLimit = input.maxLimit;
  expect(createService.getFetchParams({ _limit: input._lmit }).toMatch(...)

})
```

Comment (jozefcipa) [reply packages/strapi/lib/core-api/__tests__/service.test.js:180]: I see 😅 alright, will update it :)

Comment (jozefcipa) [reply packages/strapi/lib/core-api/__tests__/service.test.js]: Have a look now, I've also updated descriptions so it's more clear now :) 😇

Comment (jozefcipa) [reply packages/strapi/lib/core-api/__tests__/service.test.js]: image

**Review by alexandrebodin — APPROVED** [review]
LGTM, thank you fro this contribution

**Comment by alexandrebodin** [comment]
@derrickmehaffy before merging I need your ok on the doc ;)

**Comment by derrickmehaffy** [comment]
Docs look good, pending release to merge the doc PR and release.

**Review by derrickmehaffy — APPROVED** [review]

**Comment by pianomansam** [comment]
Does this apply to GraphQL? It doesn't seem to for me.

**Comment by naivedeveloper95** [comment]
> Only other thing is we need documentation for this.

Is this issue fixed? If yes, please add a code sample to demonstrate the same. If no, any workaround that I can use?

**Comment by MattieBelt** [comment]
HI @naivedeveloper95,

This PR is merged and so the issue is fixed and closed ☝🏼.
The PR for the docs is linked above: https://github.com/strapi/documentation/pull/55.
You can check out the config (with an example) here: [Strapi documentation](https://strapi.io/documentation/developer-docs/latest/setup-deployment-guides/configurations.html#api)
As far as I read, this only works with REST, not sure if it will work for GraphQL ([Comment #4931](https://github.com/strapi/strapi/issues/4931#issuecomment-874044169)), you need to test that out yourselves. 
If GraphQL doesn't work, you ofc could use the REST endpoint for this specific request.

**Pull request merged**
